network-mux/src/Network/Mux/Bearer/Socket.hs

```diff
@@ -73,10 +74,10 @@ socketAsMuxBearer tracer sd = do
                   return (header {Mx.msBlob = blob}, ts)
 
       recvLen' :: Bool -> Int64 -> [BL.ByteString] -> IO BL.ByteString
-      recvLen' _ 0 bufs = return (BL.concat $ reverse bufs)
-      recvLen' waitingOnNxtHeader l bufs = do
+      recvLen' _ 0 !bufs = return (BL.concat $ reverse bufs)
+      recvLen' waitingOnNxtHeader l !bufs = do
```

Comment:
Why does the bufs list need the `!` here? Every call for it already passes a value that's in WHNF.

Reply:
Sure they are in WHNF, but this avoids wrapping them in thunks:

```
get :: Int -> IO [String]
get = go []
  where
    go xs 0 = pure $ xs
    go xs n = do
      x <- getLine
      go (x : xs) (pred n)
```

```
> x <- get 3
a
b
c
> seq x ()
> :sprint x ()
x = _ : _
```

But
```
get :: Int -> IO [String]
get = go []
  where
    go !xs 0 = pure $ xs
    go !xs n = do
      !x <- getLine
      go (x : xs) (pred n)
```

```
> x <- get 3
a
b
c
> seq x ()
> :sprint x
x = [('a' : _),('b' : _),('c' : _)]
```